pkg/snapstore/abs_snapstore.go

```diff
@@ -51,56 +52,64 @@ type ABSSnapStore struct {
 }
 
 type absCredentials struct {
-	BucketName     string `json:"bucketName"`
-	SecretKey      string `json:"storageKey"`
-	StorageAccount string `json:"storageAccount"`
+	BucketName         string  `json:"bucketName"`
+	AccountKey         string  `json:"storageKey"`
+	AccountName        string  `json:"storageAccount"`
+	StorageAPIEndpoint *string `json:"storageAPIEndpoint,omitempty"`
 }
 
 // NewABSSnapStore create new ABSSnapStore from shared configuration with specified bucket
 func NewABSSnapStore(config *brtypes.SnapstoreConfig) (*ABSSnapStore, error) {
-	storageAccount, storageKey, err := getCredentials(getEnvPrefixString(config.IsSource))
+	absCreds, err := getCredentials(getEnvPrefixString(config.IsSource))
 	if err != nil {
 		return nil, err
 	}
 
-	credentials, err := azblob.NewSharedKeyCredential(storageAccount, storageKey)
+	sharedKeyCredential, err := azblob.NewSharedKeyCredential(absCreds.AccountName, absCreds.AccountKey)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create shared key credentials: %v", err)
 	}
 
-	p := azblob.NewPipeline(credentials, azblob.PipelineOptions{
+	p := azblob.NewPipeline(sharedKeyCredential, azblob.PipelineOptions{
 		Retry: azblob.RetryOptions{
 			TryTimeout: downloadTimeout,
 		}})
-	u, err := url.Parse(fmt.Sprintf("https://%s.%s", storageAccount, brtypes.AzureBlobStorageHostName))
+
+	storageEndpoint := fmt.Sprintf("https://%s.%s", absCreds.AccountName, brtypes.AzureBlobStorageHostName)
+	// Consumer provided endpoint overrides the default endpoint for Azure Blob Storage
+	if absCreds.StorageAPIEndpoint != nil {
+		storageEndpoint = *absCreds.StorageAPIEndpoint
+	}
```

Comment:
If I understand correctly this change:
- you expect the fully resolved URL and not just the storage domain. That means, you will no longer do the "stitching" between the domain and the storageAccount

Should you consider the `storageAccount` field then somewhat deprecated and mark it as such. The behavior with the endpoint is way too similar and it would make sense to have a "unified" way to do that in the future.

Reply:
Thanks for your feedback @kon-angelo.

Yes, you are right that the override, which is stored in `absCreds.StorageAPIEndpoint` would have the full qualified URL for the Azure resource, for example:

`https://storeageaccount.blob.core.windows.net` or
`https://storeageaccount.blob.core.chinacloudapi.cn`

"Stitching" between the domain and storage account is not performed, *only* for the override. The stitch is performed for storage accounts which have the domain as `blob.core.windows.net`.

@unmarshall @shreyas-s-rao would you like to chime in on whether we should deprecate the `storageAccount` field? I agree that it would be more "unified" but puts the burden on the consumers and operators of etcd-backup-restore to now fully form the URL, and this would make the API somewhat different just for Azure, when compared to S3 and GCS.

Reply:
@shreyas-s-rao I understand that you can't just go ahead and delete the field. I was only hinting that you mark it as deprecated (and remove it sometime in the future potentially). From the provider-extension standpoint, it makes sense that we use one option, so I believe that the storageAccount usage will naturally fade with gardener. This is just an optimization I guess

Reply:
Actually, scratch that. You always need the `accountName` now because of the way the credentials are constructed:
```
storageEndpoint := fmt.Sprintf("https://%s.%s", absCreds.AccountName, brtypes.AzureBlobStorageHostName)
```

Hmm, not sure if that makes me more inclined to think that passing just the domain part is better but 🤷

Reply:
>Hmm, not sure if that makes me more inclined to think that passing just the domain part is better but 🤷

In general you are right. The consumer should not be bothered with how the service URL is constructed and that should be responsibility of the provider implementation inside etcd-backup-restore. Only the well-defined and known constituents which are required to construct the serviceURL should be part of the API.

Ideally the service endpoint URL should be the same across using an emulator and actual blob storage. If you see [this](https://github.com/Azure/Azurite?tab=readme-ov-file#differences-between-azurite-and-azure-storage) then you will realize that azurite supports 2 different serviceURL formats. The easy one was chosen when support for an emulator was introduced but then the cost of using this option was not very clearly understood. This is now clear - it causes inconsistency in the API if both needs to be supported.

Reply:
Please check [b154104](https://github.com/gardener/etcd-backup-restore/pull/756/commits/b15410461787025718d22ad3ca4823a6ca012970).